torch/_higher_order_ops/templated_attention.py

```diff
@@ -220,3 +249,383 @@ def templated_attention_fake_tensor_mode(
+                fw_out = score_mod(*args)
+                out_requires_grad = fw_out.requires_grad
+                return ((fw_out,), (out_requires_grad,))
+
+            joint = create_joint(fw_with_masks, aot_config=dummy_aot_config)
+            args = [score, b, h, m, n] + list(other_buffers)
+            optional_grad = [example_grad] if example_grad.requires_grad else []
+            _, grads = joint(args, optional_grad)
+
+            return grads
+
+        joint_graph = make_fx(joint_f)(
+            *unwrapped_score_mod_indexes, example_grad, *unwrapped_other_buffers
+        )
+        return score_mod, joint_graph
+
+
+class TemplatedAttentionAutogradOp(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx, query, key, value, fw_graph, joint_graph, *other_buffers
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        any_buffer_requires_grad = any(buffer.requires_grad for buffer in other_buffers)
+        assert (
+            not any_buffer_requires_grad
+        ), "Captured buffers that require grad are not yet supported."
+        ctx._fw_graph = fw_graph
+        ctx._joint_graph = joint_graph
+        with torch._C._AutoDispatchBelowAutograd():
+            out, logsumexp = templated_attention(
+                query, key, value, fw_graph, *other_buffers
+            )
+
+        ctx.save_for_backward(query, key, value, out, logsumexp, *other_buffers)
+        return out, logsumexp
+
+    @staticmethod
+    def backward(ctx, grad_out, logsumexp_grad):
+        fw_args = ctx.saved_tensors
+        query, key, value, out, logsumexp, *other_buffers = fw_args
+        fw_graph = ctx._fw_graph
+        joint_graph = ctx._joint_graph
+        # We have asserted that other_buffers do not require grad in the forward
+        none_grads = [None] * (2 + len(other_buffers))
+        grad_query, grad_key, grad_value = templated_attention_backward(
+            query,
+            key,
+            value,
+            out,
+            logsumexp,
+            grad_out,
+            fw_graph,
+            joint_graph,
+            *other_buffers,
+        )
+        return grad_query, grad_key, grad_value, *none_grads
+
+
+@templated_attention.py_impl(DispatchKey.Autograd)
+def templated_attention_autograd(
+    query: torch.Tensor,
+    key: torch.Tensor,
+    value: torch.Tensor,
+    score_mod: Callable,
+    *other_buffers: Tuple[torch.Tensor, ...],
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    input_requires_grad = any(t.requires_grad for t in (query, key, value))
+    if torch.is_grad_enabled() and input_requires_grad:
+        example_vals = [
+            torch.zeros((), dtype=query.dtype, requires_grad=input_requires_grad)
+        ] + [torch.zeros((), dtype=torch.int) for _ in range(4)]
+        fw_graph, bw_graph = create_fw_bw_graph(score_mod, example_vals, other_buffers)
+    else:
+        fw_graph, bw_graph = score_mod, None
+    out, logsumexp = TemplatedAttentionAutogradOp.apply(
```

Comment:
I think we should just not even have an autograd.Function if it doesn't require grad?